### DISCUSSION TITLE
What patterns are being used for event handlers...

### DISCUSSION THREAD

**Discussion by pfw** [unpoly/unpoly#321] [Ideas]
We have an app rather than website - distinction without too much meaning but essentially lots of data tables and filtering those tables and lots of layers, etc - and one pattern we have attaching custom js event handlers with up.on within up.compiler to pick up the elements as they come in. This works perfectly from a technical point of view but is becoming a little cumbersome to keep track off and is missing a bit of discoverability - how do you know an handler is set when looking at the html?

I'm messing around with macros to try and get some info into the html so it is easier to see what is happening and can do with something like this:

``` javascript

up.macro("[x-on]", (element) => {
  up.on(element, 'click', (event) => {
    this[element.getAttribute('x-handler')](event);
  })
});

```

``` html
<button type="button" x-on="click" x-handler="optionalOpener">Open optional panel</button>
```

Rather than this:

``` javascript
up.compiler('button.optional-opener', (button) => {
  up.on(button, 'click', (event) => {
    // open the optional panel and other things that are more complicated
  }
}
```

``` html
<button type="button" class="optional-opener">Open optional panel</button>
```

The macro works in my experiments on obviously simple cases and it's nice to see in the html what is triggered and to be able the html structure without having to change the JS (change button to a, again simple case but you get the idea).

How are others approaching this? 

Do you have any of the same concerns about discoverability? 

I'm not advocating for anything particular at this point, just interested in what others are doing and if anyone else has experimented with other ways of dealing with this.

Peter W.

**Comment by triskweline** [comment]
> Do you have any of the same concerns about discoverability?

There are different schools of thought here.

Your concerns about discoverability resonates with what others have called [locality of behavior](https://htmx.org/essays/locality-of-behaviour/).

The opposite idea is that of fully unobtrusive JavaScript, where the HTML should work on its own and not care about JavaScript enhancement.

> How are others approaching this?

Personally I'm a fan of maximum unobtrusiveness, but many of my colleagues use custom attributes for behavior and classes for styling. So if you see `<button optional-opener>` in the code, you know there's a file `compilers/optional-opener.js` with the behavior.

This pattern improves discoverability, but leads to extra HTML when a component has both styles and behavior, e.g. `<div class="tabs" tabs>`.

**Reply by pfw** [reply]
Thanks Henning. 

I’m a bit torn really. I do like the unobtrusiveness generally. I wonder if another way to look at this is for the compiler/macro to ‘flag’ the element in some way so that when looking into dev tools if gives a hint that a particular element has behaviour attached. Potentially that and something like that the structure of files you mention might help. 

Thanks for your thoughts.
